Added Tex Language Support

isaacST08 opened the pull request:
Added support for the tex language. 
Added the texlab LSP along with config options for build/compiling:
- Compiling package option
- Compiling executable option
- Compiling executable args option
- PDF preview using your PDF viewer of choice
- PDF preview update on save option

Not implemented (yet):
- Support for other LSPs
- Formatting support via nvf options
- Texlab options (There is however an `extraLuaSettings` option for texlab that allows setting these options in the meantime): 
  - chktex  
  -  diagnosticsDelay
  -  diagnostics
  - symbols
  - formatterLineLength
  - bibtexFormatter
  - latexFormatter
  - latexindent
  - completion
  - inlayHints
  - experimental


Defaults mostly follow the [texlab defaults](https://github.com/latex-lsp/texlab/wiki/Configuration) with some exceptions such as the default compiler being tectonic and the default file viewer being okular.

(I also alphabetically sorted the language imports in the modules/plugins/languages/default.nix)

isaacST08 commented:
This should satisfy #525

NotAShelf commented on flake.nix:
We normally try to avoid forks for packages, though there isn't a concrete security policy against them. For some added security, you could pin the hash of your latest revision and that'd safeguard the lock from being mutated from force pushes as far as I know.

isaacST08 commented:
Feature essentially done.

Implemented:
- Lsp
  - Texlab (Recommended)
    - All features currently provided as documented by texlab
- Builders/Compilers
  - Tectonic (Recommended)
  - latexmk
- PDF Viewers
  - okular
  - zathura
  - sioyek
  - qpdfview
- Forward search
- Formatter

Not implemented:
- Inverse search
  - Not planned as widely different between pdf viewers and configuring through nix might not even be possible for some (at least not easily) (okular...).
  - Sioyek does have inverse search support via cli args and (should) be working properly with default values.


Due to the scale that this language implementation ended up being I organized it differently compared to the other languages; it is in multiple files and modules, all called by a `default.nix` in a `tex` directory. The implementation of the second largest language is more that 5 times smaller and I felt like a single 1500 line file wasn't ideal and the current implementation was a more appropriate approach.


I have not tested every single combination of options as that would take forever. I have tested many of the more relevant ones but I have not tested the texlab experimental settings for instance.

The config I am currently using is as follows:
```
settings.vim.languages.tex = {
  enable = true;
  lsp.texlab = {
    forwardSearch.enable = true;
  };
  build = {
    builders.tectonic.enable = true;
    forwardSearchAfter = true;
    onSave = true;
  };
  pdfViewer.zathura.enable = true;
  extraOpts.texFlavor = {
    enable = true;
    flavor = "tex";
  };
};
```

I have created frameworks for easily adding both new pdf viewers and new tex builders/compilers. Adding other LSP servers should be not *too* complicated if texlab proves to be fundamentally inadequate.


Overall, this implementation will cover all of my needs for tex and includes many other features I will likely not use but have been implemented to achieve near feature parity with many of the implemented programs which will hopefully cover most of what anyone else requires. If not, in a few places I added methods to be able to add custom configuration just in case.

There are no other features I plan to add for the time being, but I will try to fix bugs as they come up on my side.

NotAShelf commented:
Looks good from a first glance. Thank you for your hard work.

I'm too tired to review in full at the moment, but I'll properly review tomorrow when I wake up. One thing that comes to mid is that `mkEnableTreesitterOption` could probably be put into the extended library, so that other functions can use them as well.

NotAShelf commented:
If you could also fix the Cı errors by then, that'd be great.

isaacST08 commented:
Sounds good, I’ll get onto those things when I next get a moment

isaacST08 commented:
I made some of the changes you requested.

I made a mkEnableTreesitterOption function in the extended library. I must say I'm not too familiar with extending lib so it's a little different from how the original function was. I didn't want to start messing with the imports of the existing lib extensions by finding a way to add config as an input so the function itself requires passing the config to it as one of its inputs. If there's a better/cleaner way to do this feel free to make those changes or provide a source where I can learn how to do it properly.

If there are any other changes you wish to be made don't hesitate to let me know.


On an aside, I have been testing this PR this whole time for academic work and I can now also say that as I have now started using bibtex I can say that it has been working flawlessly and multi-file build and re-building has been working great too.

NotAShelf commented:
I'm going to have to ask you to rebase.

With #605 we have transitioned to an alternative method that is faster and cheaper than flake inputs for plugins, which does meant that there will be merge conflicts I will need you to fix before I can merge this.

Meanwhile, I'll provide a complete review, there shouldn't be too much to cover this time around.

NotAShelf commented on modules/plugins/languages/tex/build/builders/tectonic.nix:
```diff
@@ -0,0 +1,203 @@
+        if builderCfg.synctex
+        then ["--synctex"]
+        else []
+      )
+      ++ (
+        if builderCfg.untrustedInput
+        then ["--untrusted"]
+        else []
+      )
+      # Options
+      ++ (
+        if builderCfg.reruns > 0
+        then ["--reruns" "${toString builderCfg.reruns}"]
+        else []
+      )
+      ++ (
+        if builderCfg.bundle != ""
+        then ["--bundle" "${toString builderCfg.bundle}"]
+        else []
+      )
+      ++ (
+        if builderCfg.webBundle != ""
+        then ["--web-bundle" "${toString builderCfg.webBundle}"]
+        else []
+      )
+      ++ (
+        if builderCfg.outfmt != ""
+        then ["--outfmt" "${toString builderCfg.outfmt}"]
+        else []
+      )
+      ++ (concatLists (map (x: ["--hide" x]) builderCfg.hidePaths))
+      ++ (
+        if builderCfg.format != ""
+        then ["--format" "${toString builderCfg.format}"]
+        else []
+      )
+      ++ (
+        if builderCfg.color != ""
+        then ["--color" "${toString builderCfg.color}"]
+        else []
```
[`lib.optionals`](https://noogle.dev/f/lib/optionals) should be able to streamline all of this.

isaacST08 replied:
I was unaware that there was a lib function that does exactly what is needed here, thank you for the tip

NotAShelf commented on modules/plugins/languages/tex/build/default.nix:
```diff
@@ -0,0 +1,113 @@
+
+  options.vim.languages.tex.build = {
+    enable =
+      mkBool (
+        if enabledBuildersCount > 1
+        then throw "nvf-tex-language does not support having more than 1 builders enabled!"
+        else (enabledBuildersCount == 1)
+      ) ''
+        Whether to enable configuring the builder.
+
+        By enabling any of the builders, this option will be automatically set.
+        If you enable more than one builder then an error will be thrown.
+      '';
```
I'd rather if you asserted this in `config.assertions` instead of this throw here.

NotAShelf commented on modules/plugins/languages/tex/build/default.nix:
```diff
@@ -0,0 +1,113 @@
+    onSave = mkBool false "Set this property to true if you want to compile the project after saving a file.";
+
+    useFileList = mkBool false ''
+      When set to true, the server will use the .fls files produced by the TeX engine as an additional input for the project detection.
+
+      Note that enabling this property might have an impact on performance.
```
```suggestion
      When set to `true`, the server will use the `.fls` files produced by the TeX engine as an additional input for the project detection.

      Note that enabling this property might have an impact on performance.
```

NotAShelf commented on modules/plugins/languages/tex/build/default.nix:
```diff
@@ -0,0 +1,113 @@
+    filename = mkOption {
+      type = nullOr str;
+      default = null;
+      description = ''
+        Allows overriding the default file name of the build artifact. This setting is used to find the correct PDF file to open during forward search.
+      '';
```
Try to limit line length to 80 characters starting from the end of indent. Not a hard-limit, but a good rule of thumb to keep it below *at most* 110 chars.

NotAShelf commented on modules/plugins/languages/tex/default.nix:
```diff
@@ -0,0 +1,68 @@
+          default = false;
+          example = true;
+          description = ''
+            Whether to set the vim.g.tex_flavor (g:tex_flavor) option in your lua config.
```
```suggestion
            Whether to set the `vim.g.tex_flavor` (`g:tex_flavor`) option in your Lua config.
```

NotAShelf commented on modules/plugins/languages/tex/build/builders/tectonic.nix:
```diff
@@ -0,0 +1,203 @@
+      };
+
+      color = mkOption {
+        type = enum [
+          "always"
+          "auto"
+          "never"
+          ""
+        ];
+        default = "";
```
Can this be `nullOr` enum? `!= null` is slightly more contextful.

NotAShelf commented on modules/plugins/languages/tex/build/builders/tectonic.nix:
```diff
@@ -0,0 +1,203 @@
+    options = {
+      enable = mkEnableOption "Whether to enable Tex Compilation Via Tectonic";
+
+      package = mkOption {
+        type = package;
+        default = pkgs.tectonic;
+        description = "tectonic package";
+      };
```
[`mkPackageOption`](https://noogle.dev/f/lib/mkPackageOption) can be used here.

NotAShelf commented on modules/plugins/languages/tex/build/builders/tectonic.nix:
```diff
@@ -0,0 +1,203 @@
+    inherit name moduleInheritancePackage;
+
+    options = {
+      enable = mkEnableOption "Whether to enable Tex Compilation Via Tectonic";
```
```suggestion
      enable = mkEnableOption "Tex Compilation Via Tectonic";
```
`mkEnableOtion` prepends "Whether to enable..." to the description already.

NotAShelf commented on modules/plugins/languages/tex/build/builders/latexmk.nix:
```diff
@@ -0,0 +1,57 @@
+      };
+    };
+
+    args = builderCfg: (
+      # Flags
+      (
+        if builderCfg.pdfOutput
+        then ["-pdf"]
+        else []
+      )
+      # Base args
+      ++ [
+        "-quiet"
+        "%f"
+      ]
```
Put base args first, unless the ordering matters. In addition, use `lib.optionals` here.

isaacST08 replied:
Ordering does matter unfortunately, I'll make a comment of that in file stating that unless you state otherwise.

Again as above, thank you for the `lib.optionals` tip.

NotAShelf commented on modules/plugins/languages/tex/build/builders/latexmk.nix:
```diff
@@ -0,0 +1,57 @@
+    inherit name moduleInheritancePackage;
+
+    options = {
+      enable = mkEnableOption "Whether to enable Tex Compilation Via latexmk";
```
```suggestion
      enable = mkEnableOption "Tex Compilation Via latexmk";
```

NotAShelf commented on lib/types/languages.nix:
```diff
@@ -32,6 +32,13 @@
       default = ["vimPlugins" "nvim-treesitter" "builtGrammars" grammar];
     };
+
+  mkEnableTreesitterOption = config: language:
+    mkOption {
+      type = bool;
+      default = config.vim.languages.enableTreesitter;
+      description = "Whether to enable ${language} treesitter";
+    };
```
You could probably take a bool instead of the entire `config` attribute, should be cheaper. Perhaps something like

```nix
mkEnableTreesitterOption = language: condition:
    mkOption {
      type = bool;
      default = condition;
      description = "Whether to enable ${language} Treesitter";
    };
```

and used as, for example

```nix 
mkEnableTreesitterOption "lua" config.vim.languages.enableTreesitter
```

isaacST08 replied:
I made the change but implemented it so that the boolean is provided first so it can be overloaded and used with the boolean pre-filled

NotAShelf commented on modules/plugins/languages/tex/build/builders/tectonic.nix:
```diff
@@ -0,0 +1,203 @@
+        description = "The executable name from the build package that will be used to build/compile the tex.";
+      };
+
+      # -- Flags --
+      keepIntermediates = mkBool false ''
+        Keep the intermediate files generated during processing.
+
+        If texlab is reporting build errors when there shouldn't be, disable this option.
+      '';
+      keepLogs = mkBool true ''
+        Keep the log files generated during processing.
+
+        Without the keepLogs flag, texlab won't be able to report compilation warnings.
+      '';
+      onlyCached = mkBool false "Use only resource files cached locally";
+      synctex = mkBool true "Generate SyncTeX data";
+      untrustedInput = mkBool false "Input is untrusted -- disable all known-insecure features";
+
+      # -- Options --
+      reruns = mkOption {
+        type = ints.unsigned;
+        default = 0;
+        example = 2;
+        description = ''
+          Rerun the TeX engine exactly this many times after the first.
+
+          Setting this value to 0 will disable setting this option.
+        '';
+      };
```
It's better to prefix those with "Whether to ..." as boolean options. The resulting descriptions in the manual might be confusing as they are assertive rather than descriptive.

NotAShelf commented on modules/plugins/languages/tex/build/builders/tectonic.nix:
```diff
@@ -0,0 +1,203 @@
+        '';
+      };
+
+      bundle = mkOption {
+        type = str;
+        default = "";
+        description = "Use this directory or Zip-format bundle file to find resource files instead of the default";
```
This should be `nullOr str` to handle the empty path. `""` is not quite a directory.

In addition, it might better to startwith "The directory or zip-format bundle file..."`

NotAShelf commented:
I was wrong, that is a lot of changes. Although, I do appreciate your good work. TeX ecosystem is incredibly large and I was partially intimidated from working on TeX support myself.

isaacST08 commented:
I've done a rebase but the new-file-templates.nvim plugin is going to be an issue again for generating new tex files. Can you give me a rundown of how one might handle changing the source to the fork I made fixing the issue until the fix gets merged into the main repo?

In the meantime, I'll start working on the fixes you requested.

Just as a note, after rebasing by config became broken. The error was a big mess of nvim pluggin names from the looks of it (I can post the stacktrace error somewhere if you'd like but I don't know if here is where you want it). Removing `settings.vim.lsp.lspsaga.enable` fixed the issue. I don't know if this expected behaviour and that option is now (or will be) deprecated or if this is something to look into.

NotAShelf commented:
LSPSaga is addressed in #629, alongside many other plugins. I'll merge it somehow tomorrow if my config decides to build.

> Can you give me a rundown of how one might handle changing the source to the fork I made fixing the issue until the fix gets merged into the main repo?

`npins add --name <name> github <owner> <repo>` shoudl work, it's a little *less* declarative than a flake input but same idea.

isaacST08 commented:
@NotAShelf is there something I'm missing for this to be merged? Are there more changes you have for me? Or are you still busy with the migration and this should wait until later?

NotAShelf commented:
Sorry, I've put this aside for a while because we were trying to decide how language modules would be handled in the future. I'd like to get rid of lspconfig once Neovim 0.11 drops, but I am not exactly sure if I want to do so *yet*. We already started working on #593, but it's stale because there isn't an agreement on the schema.

This is a relatively large PR, and I quite unsure whether I will be able to handle the module myself if I merge it as we modify the language module. I'll revisit this PR in the coming days (before this weekend) if I cannot decide on a way to go for the language modules as a whole.

ndrwstn commented:
@isaacST08 any chance you can make your fork usable by a nix flake? I'm just coming to nix and so I couldn't figure out how to use an alternate github. LaTeX is essential to my workflow, so the lack of support by nvf has been problematic to me.

alfarelcynthesis commented:
> isaacST08 any chance you can make your fork usable by a nix flake? I'm just coming to nix and so I couldn't figure out how to use an alternate github. LaTeX is essential to my workflow, so the lack of support by nvf has been problematic to me.

You should be able to use a branch on someone's fork by replacing the url for nvf with `github:isaacST08/nvf/feature-language-tex`.

NotAShelf commented:
> Sorry, I've put this aside for a while because we were trying to decide how language modules would be handled in the future. I'd like to get rid of lspconfig once Neovim 0.11 drops, but I am not exactly sure if I want to do so _yet_. We already started working on #593, but it's stale because there isn't an agreement on the schema.
> 
> This is a relatively large PR, and I quite unsure whether I will be able to handle the module myself if I merge it as we modify the language module. I'll revisit this PR in the coming days (before this weekend) if I cannot decide on a way to go for the language modules as a whole.

Following up on this, we have mostly agreed on a schema. I'm not sure how feasible its implementation will be for Tex (although I would like to take a look after I deliver my due work). I'm going to review this PR one more time tomorrow, if I forget please remind me. We can probably merge this as is I've covered most, if not all little nitpicks I could find.

NotAShelf commented:
There seems to be a [few typos caught by the CI](https://github.com/NotAShelf/nvf/actions/runs/13640523465/job/38337487537?pr=569) and an [evaluation failure](https://github.com/NotAShelf/nvf/actions/runs/13640523473/job/38337488498?pr=569). Please address those before I can merge.

isaacST08 commented:
> There seems to be a [few typos caught by the CI](https://github.com/NotAShelf/nvf/actions/runs/13640523465/job/38337487537?pr=569) and an [evaluation failure](https://github.com/NotAShelf/nvf/actions/runs/13640523473/job/38337488498?pr=569). Please address those before I can merge.

I have fixed these but have done so via online GitHub so I will update on my system later today. 

I’m a little baffled as to why I was able to build successfully with the evaluation error on my machine but regardless it should be fixed now. I will still test properly when I get back to my machine though.

isaacST08 commented:
> > isaacST08 any chance you can make your fork usable by a nix flake? I'm just coming to nix and so I couldn't figure out how to use an alternate github. LaTeX is essential to my workflow, so the lack of support by nvf has been problematic to me.
> 
> You should be able to use a branch on someone's fork by replacing the url for nvf with `github:isaacST08/nvf/feature-language-tex`.

Yes this is how you would use my fork. I will keep updating my fork with any changes that come to nvf main every few days until it gets merged.

Additionally, if you run into any problems please open an issue on my fork’s GitHub page. This can either be for bugs or questions with configuration as I don’t know how to get you a branched version of the nvf appendix B docs.

NotAShelf commented:
I'm guessing somehow lazy eval skipped the faulty parts. Please feel free to ping me once you apply your changes, and I'll work on merging this. If there's any more nits, I can probably fix them myself over the weekend.

isaacST08 commented:
@NotAShelf I've applied the changes and it works on my machine (famous last words lol). Sorry for the delay, I had quite a big assignment due (in latex funnily enough). But at least I can say once again that its been pretty roughly tested and I haven't run into any bugs for a while (at least within the realm of my use cases).

NotAShelf commented:
No worries, I was caught up in meetings all day and only now found the time to check my notifications. I'll be home in roughly an hour, then I'll look at this PR again. I'd think everything is done, the CI was passing last time and if there are any other nits, I'll get them myself as I've said before.

NotAShelf reviewed (commented):
There are some stylistic choices that I don't particularly agree with here, but I think I've nitted enough. None of them are critical, and I'll get them whenever I refactor the language module myself.

One thing of note:

```nix
config = mkIf (cfg.enable && (any (x: x.enable) (attrValues cfg.lsp))) {
    vim.lsp.lspconfig.enable = true; # Enable lspconfig when any of the lsps are enabled
  };
  ```
  
This is defined explicitly in `modules/plugins/languages/tex/lsp/default.nix` but I'd argue that it should be a part of the lspconfig module. Is there a specific reason this is in languages/tex?

Other than that, there's nothing else left. Thank you for your efforts. If you could just get the CI failure (there's an infrec from the module) I'll merge this.

isaacST08 commented:
> There are some stylistic choices that I don't particularly agree with here, but I think I've nitted enough. None of them are critical, and I'll get them whenever I refactor the language module myself.
> 
> One thing of note:
> 
> ```nix
> config = mkIf (cfg.enable && (any (x: x.enable) (attrValues cfg.lsp))) {
>     vim.lsp.lspconfig.enable = true; # Enable lspconfig when any of the lsps are enabled
>   };
> ```
> 
> This is defined explicitly in `modules/plugins/languages/tex/lsp/default.nix` but I'd argue that it should be a part of the lspconfig module. Is there a specific reason this is in languages/tex?
> 
> Other than that, there's nothing else left. Thank you for your efforts. If you could just get the CI failure (there's an infrec from the module) I'll merge this.

I think that I wrote that quite early on in the project. I was still learning the code base and I wanted to stay in my own lane and modify as few other files as possible. I think it could belong in either location. However, I think putting it in the lspconfig module could make it a little less concise in terms of calling options of other modules but the exact same thing can be said for the other direction. It's your code base so inevitably it's your decision.

Again, I don't know why the CI is catching these errors but why I build it is not. I'm using the pdfViewer config options in my set up so it is getting evaluated but I don't see why the CI is getting an infrec but I'm not. I made a tweak that *might* fix it but I'm going to have to wait for a different day when I'm more awake to properly investigate an infrec, especially one that I apparently will have to solve using purely theory alone. 
If you can see a quick fix for it that I'm missing please feel free to make that change.

ndrwstn commented:
> Additionally, if you run into any problems please open an issue on my fork’s GitHub page. This can either be for bugs or questions with configuration as I don’t know how to get you a branched version of the nvf appendix B docs.

No worries; I poked around in the code until I got it to work. I did have a couple questions regarding: use and a possible bug I wanted to bring to your attention--but I don't think you have issues open on your fork? I'll save the conversation for there rather than here though.

isaacST08 commented:
> orries; I poked around in the code until I got it to work. I did have a couple questions regarding: u

Sorry, I forgot to turn issues on, they are on now though.

isaacST08 commented:
@NotAShelf Apologies for the delay but I spent some time working on some other projects and university. Though in that time I learned more about nix and refactored the pdf viewer section to fix the infrec error accordingly (getting adequate sleep now that finals are done helped too). Additionally, I figured out how to run the checks on my own system (I think) which now evaluates without errors which I'm hoping will be the same for when the CI runs its own checks.

Once again, it is ready for your review and as always let me know if anything goes awry and I'll do my best to fix it.

NotAShelf commented:
No worries, I believe some people have just elected to use your fork instead for Tex support.

We have finally completed the [first step of the LSP refactor](https://github.com/NotAShelf/nvf/pull/860) with the release of Neovim 0.11. This is a much simpler API, and uses Neovim *directly* to manage language servers instead of relying on lspconfig. One problem for now is that this API is exclusive to > 0.11, and causes breakage for stable channel (which still ships 0.10, so the language *modules* have not been migrated yet; we are waiting for 25.05 release. If you could update the PR to use the new API (which is quite simple, and should not require significant refactor) then I can just merge this PR with NixOS 25.05 release.

isaacST08 commented:
> No worries, I believe some people have just elected to use your fork instead for Tex support.
> 
> We have finally completed the [first step of the LSP refactor](https://github.com/NotAShelf/nvf/pull/860) with the release of Neovim 0.11. This is a much simpler API, and uses Neovim _directly_ to manage language servers instead of relying on lspconfig. One problem for now is that this API is exclusive to > 0.11, and causes breakage for stable channel (which still ships 0.10, so the language _modules_ have not been migrated yet; we are waiting for 25.05 release. If you could update the PR to use the new API (which is quite simple, and should not require significant refactor) then I can just merge this PR with NixOS 25.05 release.

I will do this switch when we get closer to the NixOS 25.05 release. I think a mini manual outlining how one might go about switching to the new api and guidelines on what you would like to see would be very helpful. If something like this already exists please link it; I took a brief search but couldn't find anything.

Since I know that at least one person is using my fork to gain TeX support I will leave it using the current implementation until Neovim 0.11 becomes part of the stable channel.

NotAShelf commented:
There is no guideline yet, but I'll add it once I sit down to revise the documentation. We're not in a rush anyway.

ItsLiyua commented:
Typst looks quite attractive. I might just do that since I plan on making my own template anyway.

horriblename commented:
sorry for the mess, due to the diff being completely messed up and hard to read, I rebased your PR onto main

horriblename commented on modules/plugins/formatter/conform-nvim/builders/builderTemplate.nix:
```diff
@@ -0,0 +1,66 @@
+      args = args cfg.build.builders.${name};
+    };
+  };
+}
```
this template isn't really helping, it's harder to read than just inlining all this, and the comments are already add more lines than it reduces lines of code

horriblename commented on modules/plugins/formatter/conform-nvim/builders/default.nix:
```diff
@@ -0,0 +1,70 @@
+    ./latexmk.nix
+    ./tectonic.nix
+  ];
+
```
as an example, this is how formatter configs are defined in [go.nix](https://github.com/NotAShelf/nvf/blob/09470524a214ed26633ddc2b6ec0c9bf31a8b909/modules/plugins/languages/go.nix?plain=1#L62-L73). Note that users don't add more options to modify our provided config, rather users should directly modify `vim.formatter.conform-nvim.formatter.append_args` for example
